service/matching/taskqueuemonitor.go

```diff
@@ -0,0 +1,116 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package matching
+
+import (
+	"context"
+	"math"
+	"time"
+)
+
+type (
+	tqmonStop    struct{}
+	tqmonGC      struct{}
+	tqmonUsurped struct{ id *taskQueueID }
+
+	// tqmonAccess is the interface by which a taskQueueMonitor observes and
+	// updates a set of taskQueueManager instances.
+	tqmonAccess interface {
+		getTaskQueues(int) []taskQueueManager
+		unloadTaskQueue(*taskQueueID)
+	}
+
+	// taskQueueMonitor performs garbage collection and unloads instances of
+	// taskQueueManager accessed via a tqmonAccess.
+	taskQueueMonitor struct {
+		sigs chan interface{}
+	}
+)
+
+// newTaskQueueMonitor constructs a new instance of taskQueueMonitor.
+func newTaskQueueMonitor() *taskQueueMonitor {
+	return &taskQueueMonitor{sigs: make(chan interface{})}
```

Comment:
`sigs: make(chan interface{})` should this be a buffer chan? or sender can be blocked?

Reply:
e.g. `tqmonStop` is received then a caller try to send another signal?

Reply:
or busy doing `m.runGC(a)`

Reply:
I prefer synchronous channels where possible and here it's possible. This is a rendezvous primitive for handing work over to another goro, not a buffer.

Reply:
>This is a rendezvous primitive for handing work over to another goro, not a buffer.

that is fine, just guarantee caller will never be blocked (e.g. no one listening on the receiving end)

Reply:
This is not a guarantee any Go channel can make unless the population is known a priori.

Reply:
let us talk, i believe there should be a way

Reply:
To be clear about what you're asking ... if we have 10,000 taskQueueManagers running and they all lose their lease at the same time you want all 10,000 to be able to signal this monitor without blocking?

Reply:
the idea is to prevent deadlock.

since the channel in question is not buffered, then sender can be blocked if there is no receiver listening on the channel.

so the question is will there be possibility to trigger a deadlock?
any contract to guarantee there is no deadlock?

what if task queue monitor is stopped but a thread call `Signal`?

Reply:
Buffered channels block too.

A deadlock is something different and I'm not aware of any language-level constructs that can prove liveness. Some systems like TLA+ can do that for some problem domains.

If the taskQueueMonitor is stopped and something calls Signal that second call blocks - I think you already know that. You're bringing into question the validity of lifecycle functions (Start/Stop) on any of our types. Is your position that all member functions must not block _even if_ the object has been used incorrectly (e.g. Start() has not been called)?

Reply:
>Buffered channels block too.

agreed, but there should be ways to guarantee not blocking on sending part of channel forever. <- is what i want.

> Is your position that all member functions must not block even if the object has been used incorrectly (e.g. Start() has not been called)?

if possible, this will reduce number of issues we will encounter in production.